### PR TITLE
list clusters by subscription/location/resourceGroup

### PR DESCRIPTION
**Pull request by katherinelc321**
### What this PR does
**Before this PR:**
We're unable to list clusters by subscription, location, resourceGroups because the functions are not set to hand the operations properly
**After this PR:**
The frontend queries ocm for clusters by filtering by subscription, location, resourceGroups
Jira: 
https://issues.redhat.com/browse/ARO-8126
https://issues.redhat.com/browse/ARO-8128
Link to demo recording: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Deployment: The deployment process was considered and addressed if required
- [ ] Testing: New code requires new unit tests.
- [ ] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [ ] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

File: frontend/pkg/frontend/ocm.go [hunk 1]
```diff
@@ -107,6 +107,79 @@ func (f *Frontend) ConvertCStoHCPOpenShiftCluster(ctx context.Context, systemDat
 	return hcpcluster, nil
 }
 
+func (f *Frontend) ConvertCStoHCPOpenShiftClusterbySubscription(cluster *cmv1.Cluster) (*api.HCPOpenShiftCluster, error) {
```
Comment (mjlshen) [inline frontend/pkg/frontend/ocm.go:110]: Potentially naive feedback, but we should use `ConvertCStoHCPOpenShiftCluster` instead - was there some issue with using it?

File: frontend/pkg/frontend/frontend.go [hunk 2]
```diff
@@ -211,7 +211,39 @@ func (f *Frontend) ArmResourceListBySubscription(writer http.ResponseWriter, req
...
+			f.logger.Error(err.Error())
+			arm.WriteInternalServerError(writer)
+			return
+		}
+		_, err = writer.Write(resp)
+		if err != nil {
+			f.logger.Error(err.Error())
+		}
+	}
+
```
Comment (mjlshen) [inline frontend/pkg/frontend/frontend.go:246]:
```suggestion
  writer.WriteHeader(http.StatusOK)
```

File: frontend/pkg/frontend/frontend.go [hunk 3]
```diff
@@ -211,7 +211,39 @@ func (f *Frontend) ArmResourceListBySubscription(writer http.ResponseWriter, req
...
+	for _, cluster := range clusters {
+		hcpCluster, err = f.ConvertCStoHCPOpenShiftClusterbySubscription(cluster)
+		if err != nil {
+			f.logger.Error(err.Error())
+			arm.WriteInternalServerError(writer)
+			return
+		}
+
+		versionedResource := versionedInterface.NewHCPOpenShiftCluster(hcpCluster)
+		resp, err := json.MarshalIndent(versionedResource, "", "    ")
```
Comment (mjlshen) [inline frontend/pkg/frontend/frontend.go:235]: What's the purpose of `json.MarshalIndent`? I believe this is unique compared to the other `ArmResource*` functions, so asking out of curiosity

**Review by mjlshen — DISMISSED** [review]

File: frontend/pkg/frontend/frontend.go (hunk 3 above)
Comment (katherinelc321) [reply frontend/pkg/frontend/frontend.go:235]: That indents the output by line instead of a big paragraph

File: frontend/pkg/frontend/frontend.go (hunk 2 above)
Comment (katherinelc321) [reply frontend/pkg/frontend/frontend.go:246]: Kept getting `http: superfluous response.WriteHeader call` with that line

File: frontend/pkg/frontend/frontend.go (hunk 3 above)
Comment (mbarnes) [reply frontend/pkg/frontend/frontend.go:235]: We should centralize JSON marshalling in the frontend for consistency.  Like, add a `MarshalJSON` function instead of calling `json.Marshal` or `json.MarshalIndent` directly.  Not sure if there's an ARM requirement for or against pretty-printing response bodies, but either way it should be determined in one place.

File: frontend/pkg/frontend/ocm.go (hunk 1 above)
Comment (mbarnes) [reply frontend/pkg/frontend/ocm.go:110]: Suggest just passing `&arm.SystemData{}` to `ConvertCStoHCPOpenShiftCluster` for now, and revisit this after https://issues.redhat.com/browse/ARO-8002 is resolved.

File: frontend/pkg/frontend/ocm.go
Comment (mjlshen) [inline frontend/pkg/frontend/ocm.go:1]: We'll want to revert the changes here - the ID/Name we return should be the Azure Resource ID/Name

Comment (katherinelc321) [reply frontend/pkg/frontend/ocm.go:1]: The resource ID is parsed from the URL and when we list clusters by subscriptions the URL doesn't have the cluster IDs

Comment (mjlshen) [reply frontend/pkg/frontend/ocm.go:1]: Gotcha, that is a non-trivial problem for us. The function as-is (before modifications) works as intended for getting clusters, which is why I was suggesting to revert changes.

The problem is that the ID we need to return is the Azure Resource ID, not the OCM ID:
```go
				ID:         cluster.ID(), // We need some way to return Azure Resource IDs here, not OCM IDs
				Name:       cluster.Name(), // This is ok
				Type:       "Microsoft.RedHatOpenShift/hcpOpenShiftClusters", // This is ok
```

I don't have an immediate solution in mind - currently our CosmosDB maps `Azure Resource ID --> OCM ID`, maybe we need an additional mapping for `OCM ID` --> `Azure Resource ID`? That doesn't feel great though :(

**Review by mbarnes — APPROVED** [review]
Approving with a couple items to follow up on:

1. ARO-8002 is resolved (RP will store system data in CosmosDB and merge it into CS data), so this part can be completed now.

2. Other than the query string, it looks like the three list handlers are identical.  I think this could be refactored into a general "list clusters" function, with the CS query string as an argument.

File: frontend/pkg/frontend/frontend.go [hunk 4]
```diff
@@ -147,6 +147,40 @@ func (f *Frontend) ArmResourceListBySubscription(writer http.ResponseWriter, req
...
+		resp, err := json.Marshal(versionedResource)
+		if err != nil {
+			f.logger.Error(err.Error())
+			arm.WriteInternalServerError(writer)
+			return
+		}
+		_, err = writer.Write(resp)
+		if err != nil {
+			f.logger.Error(err.Error())
+		}
```
[the remark this answers is not in the document]
Comment (katherinelc321) [reply frontend/pkg/frontend/frontend.go:182]: **Listing by subscription:**
`curl -X GET "localhost:8443/subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.RedHatOpenshift/hcpOpenShiftClusters?api-version=2024-06-10-preview"                 
{"id":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg/providers/Microsoft.RedHatOpenShift/hcpOpenShiftClusters/klin-test-name","location":"eastus","name":"klin-test-name","properties":{"provisioningState":"","spec":{"api":{"url":"","visibility":"external"},"console":{"url":""},"disableUserWorkloadMonitoring":false,"dns":{"baseDomain":"","baseDomainPrefix":"klin-aro-hcp"},"etcdEncryption":false,"externalAuth":{"enabled":false,"externalAuths":[]},"fips":false,"ingress":[{"ip":"","url":"","visibility":""}],"issuerUrl":"","network":{"hostPrefix":23,"machineCidr":"10.0.0.0/16","networkType":"OVNKubernetes","podCidr":"10.128.0.0/14","serviceCidr":"172.30.0.0/16"},"platform":{"etcdEncryptionSetId":"","managedResourceGroup":"klin-test-mrg","networkSecurityGroupId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg/providers/Microsoft.Network/networkSecurityGroups/myNSG","outboundType":"loadBalancer","subnetId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg/providers/Microsoft.Network/virtualNetworks/vnet/subnets/mySubnet"},"proxy":{"httpProxy":"","httpsProxy":"","noProxy":"","trustedCa":""},"version":{"availableUpgrades":[],"channelGroup":"stable","id":"openshift-v4.15.11"}}},"systemData":{"createdBy":"","createdByType":"","lastModifiedBy":"","lastModifiedByType":""},"tags":{},"type":"Microsoft.RedHatOpenShift/hcpOpenShiftClusters"}{"id":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg-2/providers/Microsoft.RedHatOpenShift/hcpOpenShiftClusters/klin-test-name-2","location":"eastus","name":"klin-test-name-2","properties":{"provisioningState":"","spec":{"api":{"url":"","visibility":"external"},"console":{"url":""},"disableUserWorkloadMonitoring":false,"dns":{"baseDomain":"","baseDomainPrefix":"klin-aro-hcp-2"},"etcdEncryption":false,"externalAuth":{"enabled":false,"externalAuths":[]},"fips":false,"ingress":[{"ip":"","url":"","visibility":""}],"issuerUrl":"","network":{"hostPrefix":23,"machineCidr":"10.0.0.0/16","networkType":"OVNKubernetes","podCidr":"10.128.0.0/14","serviceCidr":"172.30.0.0/16"},"platform":{"etcdEncryptionSetId":"","managedResourceGroup":"klin-test-mrg","networkSecurityGroupId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg-2/providers/Microsoft.Network/networkSecurityGroups/myNSG","outboundType":"loadBalancer","subnetId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg-2/providers/Microsoft.Network/virtualNetworks/vnet/subnets/mySubnet"},"proxy":{"httpProxy":"","httpsProxy":"","noProxy":"","trustedCa":""},"version":{"availableUpgrades":[],"channelGroup":"stable","id":"openshift-v4.15.11"}}},"systemData":{"createdBy":"","createdByType":"","lastModifiedBy":"","lastModifiedByType":""},"tags":{},"type":"Microsoft.RedHatOpenShift/hcpOpenShiftClusters"}%  `

[the remark this answers is not in the document]
Comment (mjlshen) [reply frontend/pkg/frontend/frontend.go:182]: We will need to assemble our list of clusters into an array and then pass it into this type: https://github.com/Azure/ARO-HCP/blob/262bbec6c3a6327dc03ea901d578082bec311e2e/internal/api/v20240610preview/generated/models.go#L313-L320
and write it out once instead of writing out all the clusters appended to each other

File: frontend/pkg/frontend/frontend.go [hunk 5]
```diff
@@ -147,6 +147,40 @@ func (f *Frontend) ArmResourceListBySubscription(writer http.ResponseWriter, req
...
+
+		versionedResource := versionedInterface.NewHCPOpenShiftCluster(hcpCluster)
+		resp, err := json.Marshal(versionedResource)
+		if err != nil {
+			f.logger.Error(err.Error())
+			arm.WriteInternalServerError(writer)
+			return
+		}
+		_, err = writer.Write(resp)
```
Comment (AldoFusterTurpin) [inline frontend/pkg/frontend/frontend.go:179]: Not sure, just a question off the top of my head: should we create a slice of clusters and then marshal + write the single variable containing all of them? Not sure if this returns a valid JSON or some valid JSONs concatenated, which results in not a valid JSON.

(Same comment for the other functions)

File: frontend/pkg/frontend/frontend.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbarnes) [reply frontend/pkg/frontend/frontend.go:182]: >and write it out once instead of writing out all the clusters appended to each other

We need to consider paging too, though, in case the result list is large.  [ARM defines is a bit of a protocol for this](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-pageable), and it's complex enough that I think it's worth deferring to a separate card.  ([ARO-RP is a good implementation reference](https://github.com/Azure/ARO-RP/blob/master/pkg/frontend/openshiftcluster_list.go))

As a 1st pass, though, I agree just stuff it all in the Value slice above.

**Review by AldoFusterTurpin — DISMISSED** [review]
Just a question/concern about Marshaling and writing the clusters to the writer. I might be wrong but I just want to check that. Thank you! 🙂

File: frontend/pkg/frontend/frontend.go (hunk 5 above)
Comment (katherinelc321) [reply frontend/pkg/frontend/frontend.go]: Output is looking good in the slice:
` curl -X GET "localhost:8443/subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.RedHatOpenshift/hcpOpenShiftClusters?api-version=2024-06-10-preview"
{"Value":[{"id":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg/providers/Microsoft.RedHatOpenShift/hcpOpenShiftClusters/klin-test-name","location":"eastus","name":"klin-test-name","properties":{"provisioningState":"","spec":{"api":{"url":"","visibility":"external"},"console":{"url":""},"disableUserWorkloadMonitoring":false,"dns":{"baseDomain":"","baseDomainPrefix":"klin-aro-hcp"},"etcdEncryption":false,"externalAuth":{"enabled":false,"externalAuths":[]},"fips":false,"ingress":[{"ip":"","url":"","visibility":""}],"issuerUrl":"","network":{"hostPrefix":23,"machineCidr":"10.0.0.0/16","networkType":"OVNKubernetes","podCidr":"10.128.0.0/14","serviceCidr":"172.30.0.0/16"},"platform":{"etcdEncryptionSetId":"","managedResourceGroup":"klin-test-mrg","networkSecurityGroupId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg/providers/Microsoft.Network/networkSecurityGroups/myNSG","outboundType":"loadBalancer","subnetId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg/providers/Microsoft.Network/virtualNetworks/vnet/subnets/mySubnet"},"proxy":{"httpProxy":"","httpsProxy":"","noProxy":"","trustedCa":""},"version":{"availableUpgrades":[],"channelGroup":"stable","id":"openshift-v4.15.11"}}},"systemData":{"createdBy":"","createdByType":"","lastModifiedBy":"","lastModifiedByType":""},"tags":{},"type":"Microsoft.RedHatOpenShift/hcpOpenShiftClusters"},{"id":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg-2/providers/Microsoft.RedHatOpenShift/hcpOpenShiftClusters/klin-test-name-2","location":"eastus","name":"klin-test-name-2","properties":{"provisioningState":"","spec":{"api":{"url":"","visibility":"external"},"console":{"url":""},"disableUserWorkloadMonitoring":false,"dns":{"baseDomain":"","baseDomainPrefix":"klin-aro-hcp-2"},"etcdEncryption":false,"externalAuth":{"enabled":false,"externalAuths":[]},"fips":false,"ingress":[{"ip":"","url":"","visibility":""}],"issuerUrl":"","network":{"hostPrefix":23,"machineCidr":"10.0.0.0/16","networkType":"OVNKubernetes","podCidr":"10.128.0.0/14","serviceCidr":"172.30.0.0/16"},"platform":{"etcdEncryptionSetId":"","managedResourceGroup":"klin-test-mrg","networkSecurityGroupId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg-2/providers/Microsoft.Network/networkSecurityGroups/myNSG","outboundType":"loadBalancer","subnetId":"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/klin-test-rg-2/providers/Microsoft.Network/virtualNetworks/vnet/subnets/mySubnet"},"proxy":{"httpProxy":"","httpsProxy":"","noProxy":"","trustedCa":""},"version":{"availableUpgrades":[],"channelGroup":"stable","id":"openshift-v4.15.11"}}},"systemData":{"createdBy":"","createdByType":"","lastModifiedBy":"","lastModifiedByType":""},"tags":{},"type":"Microsoft.RedHatOpenShift/hcpOpenShiftClusters"}],"NextLink":null}%                       `

**Review by mjlshen — APPROVED** [review]

File: frontend/pkg/frontend/frontend.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjlshen) [reply frontend/pkg/frontend/frontend.go]: Created https://issues.redhat.com/browse/ARO-8827 as a followup for pagination

**Review by mbarnes — APPROVED** [review]
